agent/src/main/java/com/teamscale/jacoco/agent/commit_resolution/git_properties/GitPropertiesLocatorUtils.java

```diff
@@ -85,22 +87,48 @@ public static String getRevisionFromGitProperties(
 		return null;
 	}
 
+	/**
+	 * VFS (Virtual File System) protocol is used by JBoss EAP and Wildfly. Example of an URL:
+	 * vfs:/content/helloworld.war/WEB-INF/classes
+	 */
 	private static Pair<File, Boolean> getVfsContentFolder(
 			URL jarOrClassFolderUrl) throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException {
-		// used by JBoss EAP and Wildfly. Example: vfs:/content/helloworld.war/WEB-INF/classes
-		Object virtualFile = jarOrClassFolderUrl.openConnection().getContent();
+		// we obtain the URL of a specific class file as input, e.g.,
+		// vfs:/content/helloworld.war/WEB-INF/classes
+		// Next, we try to extract the artefact URL from it, e.g., vfs:/content/helloworld.war
+		String artefactUrl = extractArtefactUrl(jarOrClassFolderUrl);
+		if (artefactUrl == null) {
+			return null;
+		}
+
+		Object virtualFile = new URL(artefactUrl).openConnection().getContent();
 		Class<?> virtualFileClass = virtualFile.getClass();
 		// obtain the physical location of the class file. It is created on demand in <jboss-installation-dir>/standalone/tmp/vfs
 		Method getPhysicalFileMethod = virtualFileClass.getMethod("getPhysicalFile");
 		File file = (File) getPhysicalFileMethod.invoke(virtualFile);
-		if (file.getParentFile() == null || file.getParentFile().getParentFile() == null) {
+		return Pair.createPair(file, false);
+	}
+
+	private static String extractArtefactUrl(URL jarOrClassFolderUrl) {
+		String url = jarOrClassFolderUrl.getPath().toLowerCase();
+		String[] pathSegments = url.split("/");
+		StringBuilder artefactUrlBuilder = new StringBuilder("vfs:");
+		int segmentIdx = 0;
+		String segment = pathSegments[segmentIdx];
+		while (segmentIdx < pathSegments.length && !org.conqat.lib.commons.string.StringUtils.endsWithOneOf(
+				segment, ".jar", ".war", ".ear", ".aar")) {
```

Comment:
I would suggest to move the second condition into the loop as an if statement that will break if the condition is fulfilled. Then you also do not need to append `pathSegments[segmentIdx]` afterwards. I also think you are currently missing an additional slash in that case, right?

Reply:
Yes, I am deliberately missing it. Why would we need a slash at the end?

Reply:
Not at the end. Between the last Segment and the Segment before that